src/coprocessor/endpoint.rs

```diff
@@ -281,81 +282,73 @@ impl<E: Engine> Endpoint<E> {
     fn async_snapshot(
         engine: &E,
         ctx: &kvrpcpb::Context,
-    ) -> impl Future<Item = E::Snap, Error = Error> {
-        let (callback, future) = tikv_util::future::paired_future_callback();
+    ) -> impl std::future::Future<Output = Result<E::Snap>> {
+        let (callback, future) = tikv_util::future::paired_std_future_callback();
         let val = engine.async_snapshot(ctx, callback);
-        future::result(val)
-            .and_then(|_| {
-                future.map_err(|cancel| KvError::from(KvErrorInner::Other(box_err!(cancel))))
-            })
-            .and_then(|(_ctx, result)| result)
-            // map engine::Error -> coprocessor::Error
-            .map_err(Error::from)
+        // make engine not cross yield point
+        async move {
+            val?; // propagate error
+            let (_ctx, result) = future
+                .map_err(|cancel| KvError::from(KvErrorInner::Other(box_err!(cancel))))
+                .await?;
```

Comment:
Why after `?` there is still a `result`?

Reply:
The `?` propagates the oneshot channel error. And the `result` is from what the callback returns:

`pub type Callback<T> = Box<dyn FnOnce((CbContext, Result<T>)) + Send>;`